### ISSUE TITLE
head/drop and last/take

### ISSUE THREAD

**Issue by jvoigtlaender** [ndmitchell/hlint#674]
Actually two issues/aspects here, but somewhat related, in that the decision about one would impact another suggestions as well.

First, there is this hint currently:
https://github.com/ndmitchell/hlint/blob/a84f0c04e417766ea46d87c2786adda77bddcdf5/data/hlint.yaml#L130

In my understanding, the `isNat` side condition is responsible for this:
```haskell
head (drop 3 list)
```
getting a suggestion to use `(!!)` instead:
```haskell
=> list !! 3
```
while this:
```haskell
head (drop n list)
```
gets no suggestion.

I would suggest the latter should also get the analogous suggestion:
```haskell
=> list !! n
```
(That is, the `isNat n` side condition in the hint description further above can go away.)

After all, `head (drop n list)` is exactly equivalent to `list !! n`. Whenever either of these expressions is defined, the other is also defined and has the same value. They are undefined on exactly the same inputs, and indeed I would argue that even in the undefined cases `list !! n` is better because it gives a more informative runtime error. Since, for `n` and `list` values for which `head (drop n list)` is undefined, the error message will be "head of empty list", whereas for the same values the error message for `list !! n` will be "index too large", which is the real problem here, even for the `head (drop n list)` expression, namely that one tries to access a list element that isn't there.

**So my first suggestion here is to remove the `isNat n` condition from the `head`/`drop` rule.**

Secondly, I have seen students doing something similar to `head`/`drop`, namely:
```haskell
last (take n list)
```
From the perspective of professional Haskell programming that is maybe surprising, but actually in teaching at a point where the construction of lists from `[]` and `(:)` has not yet been discussed, so the students do not have a specific left-leaning or right-leaning interpretation of lists yet, they have no real reason to prefer `head` over `last`. So they do write stuff like in the last snippet above.

My second suggestion is to add a rule, possibly in the teaching group, that propses to replace such expressions by
```haskell
=> list !! (n - 1)
```
And the resolution about the `isNat` side condition for `head`/`drop` might have an impact on whether the proposed `last`/`take` rule should have its own side condition or not.

**Comment by jvoigtlaender** [comment]
Concerning the error message aspect: `list !! n` will also sometimes say "negative index" where `head (drop n list)` again only says a less informativ "head of empty list".

**Comment by ndmitchell** [comment]
`head (drop (-1) "neil" == 'n'` while `"neil" !! (-1) == error` - so unfortunately I can't drop the `isNat` test - despite the fact I'd really like to. I'm ok with changing the error message, but making it suddenly fail is not ok.

As for `last/take`, that seems like a good hint - add away in the normal section, but I suspect the `isNat` is still a requirement.

**Comment by jvoigtlaender** [comment]
Ah, yeah, you are absolutely right about the negative case.

Plus, for `last`/`take` it also won't work out. Since:
```
last (take 10 "janis") == 's'
```
while:
```
"janis" !! (10-1)  ->  error, index too large
```
And here there isn't an "easy" way to add a side condition, since it would not need to have to guard against negatives, but against "too large" compared to a not generally known list length.

So, I'm closing.

**Comment by jvoigtlaender** [comment]
Actually, one possibility would be, for `head`/`drop` (and probably something similar for `take`/`last` would also be possible):
```yaml
 - warn: {lhs: head (drop n x), rhs: x !! max 0 n}
```
That would really work in all cases.

But maybe it's more something that would make sense to be added into a local hint database?

One thing I am wondering now is how that would work in tandem with the existing rule
```yaml
 - warn: {lhs: head (drop n x), rhs: x !! n, side: isNat n}
```
Of course, for not known to be non-negative inputs, only the rule with `max` would fire.

But what happens for an expression like `head (drop 5 list)`? Where the side condition `isNat` is satisfied. Would then two hints be produced, one plain `list !! 5` and the other (superfluously in this case) as well, `list !! max 0 5`?

**Comment by jvoigtlaender** [comment]
Would it be possible to formulate the rule as follows:
```yaml
 - warn: {lhs: head (drop n x), rhs: x !! max 0 n, side: not (isNat n)}
```
or, in other words, what is the language of `side` conditions? Arbitrary Boolean expressions?

**Comment by ndmitchell** [comment]
Yes, side conditions can be `not`, `&&` and `||`. I really like the `max 0 n` formulation - I'd welcome a patch for that. Not sure there's something as easy for `last`, but if you can think of it, let me know.
